File: java/src/main/java/com/kingfisher/common/dataaccess/Scope.java
package com.kingfisher.common.dataaccess;

import lombok.Data;

import java.util.List;

@Data
public class Scope {
    private String type;
    private String field;
    private Object value;
    private List<Long> ids;

    public static Scope all() {
        Scope s = new Scope();
        s.type = "all";
        return s;
    }

    public static Scope self(String field, Long userId) {
        Scope s = new Scope();
        s.type = "self";
        s.field = field;
        s.value = userId;
        return s;
    }

    public static Scope department(String field, List<Long> deptIds) {
        Scope s = new Scope();
        s.type = "department";
        s.field = field;
        s.ids = deptIds;
        return s;
    }

    public static Scope subtree(String field, List<Long> deptIds) {
        Scope s = new Scope();
        s.type = "subtree";
        s.field = field;
        s.ids = deptIds;
        return s;
    }

    public boolean isAll() { return "all".equals(type); }
    public boolean isSelf() { return "self".equals(type); }
    public boolean isDepartment() { return "department".equals(type); }
    public boolean isSubtree() { return "subtree".equals(type); }
}

File: core/dataaccess/scope.go
// Package dataaccess 提供显式的数据访问范围 GORM Scope。
package dataaccess

import (
	"fmt"
	"regexp"

	"gorm.io/gorm"
)

type Kind string

const (
	KindAll        Kind = "all"
	KindSelf       Kind = "self"
	KindDepartment Kind = "department"
	KindSubtree    Kind = "subtree"
)

type Scope struct {
	kind  Kind
	field string
	ids   []uint
	user  uint
}

func All() Scope { return Scope{kind: KindAll} }

func Self(field string, userID uint) Scope {
	return Scope{kind: KindSelf, field: field, user: userID}
}

func Department(field string, departmentIDs []uint) Scope {
	return Scope{kind: KindDepartment, field: field, ids: append([]uint(nil), departmentIDs...)}
}

func Subtree(field string, departmentIDs []uint) Scope {
	return Scope{kind: KindSubtree, field: field, ids: append([]uint(nil), departmentIDs...)}
}

func Apply(db *gorm.DB, scope Scope) *gorm.DB {
	if scope.kind == KindAll {
		return db
	}
	if !validColumn(scope.field) {
		_ = db.AddError(fmt.Errorf("data access: invalid column %q", scope.field))
		return db
	}
	switch scope.kind {
	case KindSelf:
		return db.Where(scope.field+" = ?", scope.user)
	case KindDepartment, KindSubtree:
		if len(scope.ids) == 0 {
			return db.Where("1 = 0")
		}
		return db.Where(scope.field+" IN ?", scope.ids)
	default:
		_ = db.AddError(fmt.Errorf("data access: unsupported scope %q", scope.kind))
		return db
	}
}

var validColumnPattern = regexp.MustCompile(`^[a-z][a-z0-9_]*$`)

func validColumn(column string) bool { return validColumnPattern.MatchString(column) }
